### ISSUE TITLE
Standardize Command Line Arguments across clarityinc services

### ISSUE THREAD

**Issue by cdmeister** [clarityinc/clarityeco#27]
We need to make sure that we come up with a standardize coding style for parsing command line arguments. So far from our meeting we have several options:

1. Boost Program options (C++)
2. Getopt (C/C++)
3. Standard argc/argv(Multiple Languages)

Since services are in different languages we should come up with some method/function that we can use across different languages.

**Comment by cdmeister** [comment]
What is the current status on the cmd args? Is this gonna be on the release or for the next one?

**Comment by cdmeister** [comment]
Just a note, from the meeting today, we could make a "hidden service" that can standardize our args as well.

**Comment by Lilisys** [comment]
Status Update:
I'm using the Djinn workflow as a basis. That workflow is:
- client talks to cc, receives ip and port for service.
- client talks to service (img/nlp). client sets up a QuerySpec for the service containing the cmd args pushed into tags and calls service->SubmitQuery
- service does some preprocessing, which includes parsing the tags from client. service talks to CC, receives ip and port for djinn.
- service sets up its own QuerySpec for djinn. it is key that the tags consist of request name and length while input consists of  the data to be analyzed. service calls djinn->SubmitQuery.
- djinn runs, returns output to service, which returns output to client

In this system, the client uses a function to parse argc/argv and push the values into the QuerySpec.inputSet.queryInput tags. Those tags are passed along to the service, which uses if/else statements to parse them and set flags (stored as private member variables). The reason I did this was because it built upon the existing system in tonic, that is the service using an instance TonicSuiteApp which kept track of all the flags. 

For the sirius applications, (correct me if I'm wrong) it seems like the only argument is the name/location of the input file (the data itself is not sent via tags). The client (either a command line or the mobile app) can send the filename, which gets pushed into tags, and the service can pull it out and access it accordingly. 

Since Hailong's design for tags is so flexible, each service can receive tags the same way and parse them however it wants. This means the client just has to send everything over and it's the service's job to make sense of the information. 

Also, I wasn't at the meeting, so I didn't get to hear about the hidden service idea. Could someone please elaborate on that? I feel like using a hidden service could be useful in pushing the client argc/argv onto tags, but each service would still need to parse them on its own and set its own flags accordingly (makes more sense as an internal function rather than external service to me).

**Comment by Lilisys** [comment]
It has been brought to my attention that this issue is about standardizing args for starting a server rather than for tag parsing between services. My previous comment was about the latter, so sorry about the confusion.

I'm looking into the former right now. So far the djinn server has command line parsing implemented via boost program options, but I'll look into the other methods as well.

**Comment by jhauswald** [comment]
Ok looks like you cleared up your confusion but the last part I agree with:

> I feel like using a hidden service could be useful in pushing the client argc/argv onto tags, but each service would still need to parse them on its own and set its own flags accordingly (makes more sense as an internal function rather than external service to me).

This should be in a section of common functionalities used across all the services. This will eliminate any errors from copy pasting code, or make it easy to deploy new arguments across all the services. Can you think of any other common functionalities that could go into this folder @Lilisys @cdmeister ? It's similar to how djinn has a `common/` [folder](https://github.com/jhauswald/djinn/tree/master/common).

**Comment by Lilisys** [comment]
The stuff in [gen-cpp](https://github.com/Lilisys/clarityeco/tree/mergeTonic/gen-cpp) will be common to all IPAservices. Currently, there are 3 files in there that aren't thrift-generated [ServiceTypes.h](https://github.com/Lilisys/clarityeco/blob/mergeTonic/gen-cpp/ServiceTypes.h), [commons.h](https://github.com/Lilisys/clarityeco/blob/mergeTonic/gen-cpp/commons.h), and [commons.cpp](https://github.com/Lilisys/clarityeco/blob/mergeTonic/gen-cpp/commons.cpp), so we can move those out into a `service-commons/` directory. 

In terms of command line arguments, it seems like between all services, each one only _needs_ its own port and the cc port. We could make sure that if a service is started with at least 2 arguments (such as `./service 1234 5678 --othertags`), then the first is the service port and the second is the cc port. Then if the service can accept other tags (for example, djinn), that service would need its own parse_opts function (boost program options is nice because it stores all the options in a map rather than a bunch of stray variables). We wouldn't be able to make a standard parse_opts function since each service has different needs, but we could make a common function to set the port and ccport.

However, I think boost::po is only available for c++, while something like getopt is available in several languages. But since each service is in charge of its own tags (excluding port and ccport), we might not necessarily need the same parser for each service. What do you think?

**Comment by jhauswald** [comment]
This looks good and yes expectedly it'll probably turn out that the services will have little to no overlap between them.

- Boost for C++
- ArgParse for Python

The design we should have in mind is just to be conscious that there are other services so it should be easy to look into another and sort of be familiar with the structure. I imagine this constantly growing and if it can grow in a way that it doesn't look like 10 different people wrote it, that'd be ideal. That was the idea in making sure running the services was fairly standard.

**Comment by Lilisys** [comment]
Sweet. I'll make sure that djinn, tonic-image and tonic-nlp have this structure then start some documentation.

**Comment by Lilisys** [comment]
We're requiring these yes? The existing command center has the option to use default ports/ip's, but I wanna know if I should ignore those defaults and enforce the arguments.

**Comment by jhauswald** [comment]
You should insert defaults for the parameters, similar to how djinn  does it so if you run `./cc` it should run with whatever you put as default

**Comment by Thomas-Yang** [comment]
I was thinking another way to accommodate the command line arguments is to provide a configuration file with default settings. The configuration file can be both unified and customized across services. Any thoughts?

**Comment by cdmeister** [comment]
So what happens every time a new service want to join clarityeco? Do you just add something into the configuration file?

**Comment by Thomas-Yang** [comment]
I was talking about the configuration for each service(e.g. ip, port, or something else). We can define something common across the services and allow each service to customize upon that. 
(Are we talking the same thing?)

**Comment by cdmeister** [comment]
I was assuming that the config file had configuration for each service.

**Comment by Lilisys** [comment]
@cdmeister So a config file would look something like [this](https://github.com/Lilisys/clarityeco/blob/mergeTonic/gen-cpp/ServiceTypes.h), as just a list of constants? 
@Thomas-Yang I don't think I quite understand what your described setup looks like.

**Comment by Thomas-Yang** [comment]
I was talking about something like the new command center deployed on clarity07. If you look at the configuration files under "service_project/commandcenter/conf”.

**Comment by Lilisys** [comment]
So is this a single file shared among the services? Or will each service have its own config file?
If we go with the latter case, I don't see how that's very different from just defining defaults in the constructor (existing implementation).

**Comment by Lilisys** [comment]
Actually, I don't like the idea of combining pure argv/argc with some parser. What if someone wants to just use default values but specific tags? That would break if we're assuming they must use `./service 1234 5678 123.456.789.10 --tags` but they just type `./service --tags`
Since each service will have a different parser depending on what language it's written in, I think the main thing we can standardize is the flags to be set. What if we use

--port/-p   = own port number
--ccport/-c = cc port number
--ccip/-i = cc IP address (We should be able to use -i for this rather than input because the services won't take input this way)

We can set defaults (using config files) for these flags in the parser function for each service.

**Comment by jhauswald** [comment]
I agree with you @Lilisys but a config file AND boost options is overkill. I think config files are very useful when you have many things to set or need lots of additional information, like [this one](https://github.com/numpy/numpy/blob/master/site.cfg.example) I recently came across. In the command center this may end up happening at some point, seeing at the complexity that I heard from @Thomas-Yang's research.

Two scenarios:
- we do a "for-now" and just use boost assuming the "glue" command center won't get any more complex
- make a config file for the CC assuming future complexity

From this conversation, feels as though the command center and the services should be treated separetely as they both incur their own complexities.

**Comment by Lilisys** [comment]
@VivienZhou mentioned that we should probably have the service's ip address as an argument too, so I added that in

**Comment by cdmeister** [comment]
Do we want some sort of "help menu" to assist the user on the command line? In the [wiki](https://github.com/clarityinc/clarityeco/wiki/ClarityEco-Standard-Command-Line-Arguments), if nothing is specified, it assumes default values. I think that it could potentially be helpful if something like this was in place:

 `./server --help'

the help menu would be customized per service.

**Comment by cdmeister** [comment]
For boost, you end up making a help menu anyways.
